mewarpx/mewarpx/diags_store/particle_histogram_diagnostic.py

```diff
@@ -145,37 +109,126 @@ def get_fileprefix(self, species):
+            kwargs: See :class:`mewarpx.diags_store.diag_base.WarpXDiagnostic`
+                for more timing options.
+        """
+        self.assembly = assembly
+        self.plot = plot
+
+        # sanity check
+        if not isinstance(self.assembly, assemblies.ZPlane):
+            raise AttributeError(
+                "ZplanePHistDiag should only be used with a ZPlane assembly."
+            )
+
+        # set sensible linres based on simulation geometry - can override
+        # user given arguments if they don't make sense
+        if mwxrun.dim == 1:
+            logger.warn(
+                "ZPlane particle scraping histograms in 1d are trivial so this "
+                "diagnostic will not be installed."
+            )
             return
+        elif mwxrun.dim == 2:
+            linres_y = 1
+
+        # set up domain and number of bins
+        self.domain = [(mwxrun.xmin, mwxrun.xmax), (mwxrun.ymin, mwxrun.ymax)]
+        self.linres = [linres_x, linres_y]
+
+        # for now we hard code this diagnostic to only save weight but this
+        # can be extended in the future
+        self.quantities = ['x', 'y', 'w']
+
+        super(ZplanePHistDiag, self).__init__(
+            diag_steps=diag_steps, species_list=species_list, **kwargs
+        )
 
-        if self.plot:
-            self.plot_histograms()
+    def _process_scraped_particles(self, scraped_particle_dict):
+        """Function to process the scraped particle data.
 
+        Arguments:
+            scraped_particle_dict (dict): Dictionary containing scraped
+                particle data.
+        """
         for ii, species in enumerate(self.species_list):
-            fileprefix = self.get_fileprefix(species.name)
-            np.save(fileprefix + '.npy', self.Harray[ii,...])
+            idx = np.where(
+                scraped_particle_dict['species_id'] == species.species_number
+            )
+            sample = np.zeros((np.size(idx), 2))
+            sample[:,0] = scraped_particle_dict['x'][idx]
+            sample[:,1] = scraped_particle_dict['y'][idx]
+            H, _ = np.histogramdd(
+                sample=sample, bins=self.bins,
+                weights=scraped_particle_dict['w'][idx]
+            )
+            self.Harray[ii,:,:,0] += H
+
+        # if this is a diagnostic period save the data
+        if self.check_timestep():
+            self.save_and_reset_histogram()
+
+    def save_and_reset_histogram(self):
+        """Save and reset the histogram."""
+        # sum the histograms from all the processors
+        self.Harray = parallel_util.parallelsum(self.Harray)
 
+        if mwxrun.me == 0:
+            if self.plot:
+                self.plot_histograms()
+
+            for ii, species in enumerate(self.species_list):
+                fileprefix = self.get_fileprefix(species.name)
+                np.save(fileprefix + '.npy', self.Harray[ii,...])
         self.Harray[:] = 0.0
```

Comment:
I think it would be appropriate for all this to go in the generic class, and have the subclass only pass `sample` and `weights`

Reply:
(for lines 181 - 203, if that's not clear)